<a href="https://colab.research.google.com/github/nairmeghna/projects/blob/main/Diabetes_Prediction_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# importing the dependencies

import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score

# **About Dataset**

This dataset contains diabetes information about only females. It contains the following columns:


*   Pregnancies
*   Glucose

*   BloodPressure
*   SkinThickness : The collagen content

*   Insulin
*   BMI : Body Mass Index

*   DiabetesPedigreeFunction : The liklihood of diabetes based age and family history
*   Age

*   Outcome : Shows 0 for non diabetic and 1 for diabetic observations











In [ ]:
# loading the dataset

dataset = pd.read_csv("/content/diabetes.csv")
dataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
# number of rows and columns in the dataset

dataset.shape

(768, 9)

This means that there are 768 rows (i.e. 768 observations) and 9 categories.

In [ ]:
# getting some statistical measurements

dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

Here, count shows the number of datapoints we have, mean is the average value of all the 768 observations.

In [ ]:
dataset['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

This shows us that there are 500 observations that do not have diabetes and 268 have diabetes.

In [ ]:
dataset.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

And the above code shows the mean value of each category for each outcome. Please note that for the Outcome 1, all the categories have higher values than for Outcome 0.

# **Standardizing the data**

In [ ]:
# seperating the data and labels

X = dataset.drop(columns = 'Outcome', axis = 1)
Y = dataset['Outcome']

Next, we will stardardize the data. This is done to bring all the values in all the categories in a particular range so that it is easy for the ML Model to predict data.

In [ ]:
# Standardizing the data

scaler = StandardScaler()
scaler.fit(X)

# tranforming this data

standardized_data = scaler.transform(X)

print(standardized_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


It is observed that after standardization all the values are in the same range.

In [ ]:
# defining variables again

X = standardized_data
Y = dataset['Outcome']

# **Splitting the data**

In [ ]:
# splitting data into train and test

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify = Y, random_state = 2)

In [ ]:
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


Here, we have taken 20% of the data as test data (specified by test_size = 0.2).

In [ ]:
# training the model

classifier = svm.SVC(kernel = 'linear')

In [ ]:
# training the svm classifier
classifier.fit(X_train, Y_train)

SVC(kernel='linear')

# **Model Evaluation**

In [ ]:
# model evaluation by finding the accuracy score on training data

X_train_prediction = classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print("Accuracy score of the training data: ", training_data_accuracy)

Accuracy score of the training data:  0.7866449511400652


So the accuracy score on training data is above 75% (~79%) which is pretty good. Next we will check the accuracy for test data.

In [ ]:
# model evaluation by finding the accuracy score on test data

X_test_prediction = classifier.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print("Accuracy score of the test data: ", test_data_accuracy)

Accuracy score of the test data:  0.7727272727272727


The accuracy score for test data is 77% which is very close to 79% obtained for training data. This means that the model has not encountered overfitting.

# **Building the Model**

In [ ]:
# building a predicting system

input_data = (4, 110, 92, 0, 0, 37.6, 0.191, 30)

# changing the list datatype to a numpy array

input_data_as_np = np.asarray(input_data)

# reshaping the array since we are predicting only for 1 instance

input_data_reshaped = input_data_as_np.reshape(1, -1)

# standardizing the input_data

std_data = scaler.transform(input_data_reshaped)
print(std_data)

prediction = classifier.predict(std_data)
print(prediction)

if (prediction[0] == 0):
  print('The person is not diabetic')
else:
  print('The person is diabetic')

[[ 0.04601433 -0.34096773  1.18359575 -1.28821221 -0.69289057  0.71168975
  -0.84827977 -0.27575966]]
[0]
The person is not diabetic


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


The model has correctly predicted the value of the Outcome for the standardized data that was fed into it.

In [ ]:
# building a predicting system

input_data = (5, 160, 94, 23, 102, 40, 0.6, 40)

# changing the list datatype to a numpy array

input_data_as_np = np.asarray(input_data)

# reshaping the array since we are predicting only for 1 instance

input_data_reshaped = input_data_as_np.reshape(1, -1)

# standardizing the input_data

std_data = scaler.transform(input_data_reshaped)
print(std_data)

prediction = classifier.predict(std_data)
print(prediction)

if (prediction[0] == 0):
  print('The person is not diabetic')
else:
  print('The person is diabetic')

[[0.3429808  1.22388954 1.28699125 0.15453319 0.19276481 1.01629594
  0.38694877 0.57511787]]
[1]
The person is diabetic


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
